pkg/controller/workload/jobframework/helpers.go

```diff
@@ -0,0 +1,236 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package jobframework
+
+import (
+	"context"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	kueue "sigs.k8s.io/kueue/apis/kueue/v1beta1"
+	utilpriority "sigs.k8s.io/kueue/pkg/util/priority"
+	"sigs.k8s.io/kueue/pkg/workload"
+)
+
+// EnsureOneWorkload will query for the single matched workload corresponding to job and return it.
+// If there're more than one workload, we should delete the excess ones.
+// The returned workload could be nil.
+func EnsureOneWorkload(ctx context.Context, cli client.Client, req ctrl.Request, record record.EventRecorder, job GenericJob) (*kueue.Workload, error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	// Find a matching workload first if there is one.
+	var toDelete []*kueue.Workload
+	var match *kueue.Workload
+
+	if pwName := job.ParentWorkloadName(); pwName != "" {
+		pw := kueue.Workload{}
+		NamespacedName := types.NamespacedName{
+			Name:      pwName,
+			Namespace: job.Object().GetNamespace(),
+		}
+		if err := cli.Get(ctx, NamespacedName, &pw); err != nil {
+			if !apierrors.IsNotFound(err) {
+				return nil, err
+			}
+			log.V(2).Info("job with no matching parent workload", "parent-workload", pwName)
+		} else {
+			match = &pw
+		}
+	}
+
+	var workloads kueue.WorkloadList
+	if err := cli.List(ctx, &workloads, client.InNamespace(job.Object().GetNamespace()),
+		client.MatchingFields{OwnerKey: job.Object().GetName()}); err != nil {
+		log.Error(err, "Unable to list child workloads")
+		return nil, err
+	}
+
+	for i := range workloads.Items {
+		w := &workloads.Items[i]
+		owner := metav1.GetControllerOf(w)
+		// Indexes don't work in unit tests, so we explicitly check for the
+		// owner here.
+		if owner.Name != job.Object().GetName() {
+			continue
+		}
+		if match == nil && job.EquivalentToWorkload(*w) {
+			match = w
+		} else {
+			toDelete = append(toDelete, w)
+		}
+	}
+
+	// If there is no matching workload and the job is running, suspend it.
+	if match == nil && !job.IsSuspend() {
```

Comment:
I used to think about move this out of the function then we can handle the Start & Stop operation in the main reconciler loop. But when the matched workload is nil we don't know the injected nodeSelector, so I raised the issue: https://github.com/kubernetes-sigs/kueue/issues/518

We can leave it as it-is and discuss about it in that issue.

Reply:
It seems a separate issue, leaving as-is for now.